lib/auth/methods.go

```diff
@@ -312,6 +329,83 @@ func (s *Server) authenticatePasswordless(ctx context.Context, req AuthenticateU
 	return dev, user, nil
 }
 
+func (s *Server) authenticateHeadless(ctx context.Context, req AuthenticateUserRequest) (mfa *types.MFADevice, err error) {
+	// this authentication requires two client callbacks to create a headless authentication
+	// stub and approve/deny the headless authentication, so we use a standard callback timeout.
+	ctx, cancel := context.WithTimeout(ctx, defaults.CallbackTimeout)
+	defer cancel()
+
+	// Delete the headless authentication upon failure.
+	defer func() {
+		if err != nil {
+			if err := s.DeleteHeadlessAuthentication(ctx, req.HeadlessAuthenticationID); err != nil && !trace.IsNotFound(err) {
+				log.Debugf("Failed to delete headless authentication: %v", err)
+			}
+		}
+	}()
+
+	headlessAuthn := &types.HeadlessAuthentication{
+		ResourceHeader: types.ResourceHeader{
+			Metadata: types.Metadata{
+				Name: req.HeadlessAuthenticationID,
+			},
+		},
+		User:            req.Username,
+		PublicKey:       req.PublicKey,
+		ClientIpAddress: req.ClientMetadata.RemoteAddr,
```

Comment:
Is there any other place where we can obtain an IP address? The description of that field says, "IP address can be obtained for logging purposes". This is a little bit different than "IP can be used to generate a secure certificate".

CC @AntonAM

Reply:
I think this IP address is as secure as we can reasonable achieve, but it can be spoofed with an improper networking setup that doesn't fully seal off teleport behind a load balancer.

Reference: https://github.com/gravitational/teleport/blob/master/rfd/0110-ip-pinning-and-propagation.md#security

Reply:
I've updated the comment to clarify how the IP address is derived and that it's not just for logging.